### PR TITLE
sys-kernel/installkernel-systemd-boot: add petitboot config converter

### PR DESCRIPTION
**Pull request by gyakovlev**
converts loader entries to kboot.conf

example:

_cat /boot/loader/entries/1fff4ddbdcba4430889a2e52b405d08d-5.15.3*_
```
title      Gentoo Linux
version    5.15.32-talos64
machine-id 1fff4ddbdcba4430889a2e52b405d08d
options    root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash
linux      /1fff4ddbdcba4430889a2e52b405d08d/5.15.32-talos64/linux
initrd     /1fff4ddbdcba4430889a2e52b405d08d/5.15.32-talos64/initrd
title      Gentoo Linux
version    5.15.33-talos64
machine-id 1fff4ddbdcba4430889a2e52b405d08d
options    root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash
linux      /1fff4ddbdcba4430889a2e52b405d08d/5.15.33-talos64/linux
initrd     /1fff4ddbdcba4430889a2e52b405d08d/5.15.33-talos64/initrd
title      Gentoo Linux
version    5.15.35-talos64
machine-id 1fff4ddbdcba4430889a2e52b405d08d
options    root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash
linux      /1fff4ddbdcba4430889a2e52b405d08d/5.15.35-talos64/linux
initrd     /1fff4ddbdcba4430889a2e52b405d08d/5.15.35-talos64/initrd
title      Gentoo Linux
version    5.15.37-talos64
machine-id 1fff4ddbdcba4430889a2e52b405d08d
options    root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash
linux      /1fff4ddbdcba4430889a2e52b405d08d/5.15.37-talos64/linux
initrd     /1fff4ddbdcba4430889a2e52b405d08d/5.15.37-talos64/initrd
```

will get converted to
```
default=5.15.37-talos64
5.15.37-talos64='/1fff4ddbdcba4430889a2e52b405d08d/5.15.37-talos64/linux initrd=/1fff4ddbdcba4430889a2e52b405d08d/5.15.37-talos64/initrd root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash'
5.15.35-talos64='/1fff4ddbdcba4430889a2e52b405d08d/5.15.35-talos64/linux initrd=/1fff4ddbdcba4430889a2e52b405d08d/5.15.35-talos64/initrd root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash'
5.15.33-talos64='/1fff4ddbdcba4430889a2e52b405d08d/5.15.33-talos64/linux initrd=/1fff4ddbdcba4430889a2e52b405d08d/5.15.33-talos64/initrd root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash'
5.15.32-talos64='/1fff4ddbdcba4430889a2e52b405d08d/5.15.32-talos64/linux initrd=/1fff4ddbdcba4430889a2e52b405d08d/5.15.32-talos64/initrd root=ZFS=zroot/ROOT/default earlycon=hvc0 console=hvc0 console=tty0 iommu=nobypass pci=realloc video=offb:off audit=1 loglevel=3 rd.udev.log-priority=3 systemd.show_status=1 rd.systemd.show_status=auto crashkernel=4096M bootfs.snapshot quiet splash'
```

### CODE REVIEW

File: sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-1.ebuild [hunk 1]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2019-2021 Gentoo Authors
+# Copyright 2019-2022 Gentoo Authors
```
Comment (floppym) [inline sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-1.ebuild:1]: Spurious copyright update.

File: sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install [hunk 2]
```diff
@@ -0,0 +1,66 @@
+#!/usr/bin/env bash
+
+# (c) 2019-2022 Georgy Yakovlev
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install:3]: Man, you were writing it for three years? What a dedication!

File: sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+# public-domain license
+
+find_entries() {
+	local -n arr="${1}"
+	IFS= mapfile -d '' arr < <(find /boot/loader/entries -maxdepth 1 -xtype f \
+		-iname "*.conf" -print0  2>/dev/null | sort -Vurz - -- )
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install:9]: Two spaces before `2>` intentional?

File: sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		do
+			conf[${opt}]="${arg}"
+		done < "${entry}"
+
+		# check that kernel is there indeed
+		[[ -f /boot/${conf[linux]} ]] || break
+		# if initrd arg is not empty, ensure file exists
+		if [[ ${conf[initrd]} ]]; then
+			[[ -f /boot/${conf[initrd]} ]] || break
+		fi
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install:28]: Wouldn't it be better to merge this into one condition?

File: sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+	emit=$(
+		# we got version from kernel-install
+		# so set is as default if kernel exists
+		if [[ -f /boot/${machine_id}/${2}/linux ]]; then
+			echo default="${2}"
+		fi
+		# pass by reference and echo if found/validated
+		parse_entries entries
+	)
+	[[ ${emit} ]] && echo "${emit[@]}" > /boot/kboot.conf.new
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/files/91-loaderentry2kboot.install:61]: So if emit ends up empty, we're removing the config and leaving the user without one?

File: sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-2.ebuild [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Copyright 2019-2022 Gentoo Authors
+# Distributed under the terms of the GNU General Public License v2
+
+EAPI=7
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-2.ebuild:4]: I suppose 8 is fine now.

File: sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-2.ebuild [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+S=${WORKDIR}
+
+LICENSE="public-domain"
+SLOT="0"
+KEYWORDS="amd64 arm64 ~ppc64 x86"
+IUSE="petitboot"
+
+RDEPEND="|| ( sys-apps/systemd sys-boot/systemd-boot )
+	!<sys-apps/debianutils-4.9-r1[installkernel(+)]
+	!sys-kernel/installkernel-gentoo"
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-2.ebuild:17]: Let's do:

```
RDEPEND="
    ....
"
```

File: sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-2.ebuild [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+	# we could technically use a symlink here but it would require
+	# us to know the correct path, and that implies /usr merge problems
+	into /
+	newsbin - installkernel <<-EOF
+		#!/usr/bin/env sh
+		exec kernel-install add "\${1}" "\${2}"
+	EOF
+
+	exeinto /usr/lib/kernel/install.d/
+	doexe "${FILESDIR}/00-00machineid-directory.install"
```
Comment (mgorny) [inline sys-kernel/installkernel-systemd-boot/installkernel-systemd-boot-2.ebuild:29]: BTW I really feel like it's time to "g2" a repository for it and stop bundling everything in FILESDIR.

File: profiles/arch/base/use.mask [hunk 9]
```diff
@@ -1,6 +1,11 @@
 # Copyright 1999-2022 Gentoo Authors
 # Distributed under the terms of the GNU General Public License v2
 
+# Georgy Yakovlev <gyakovlev@gentoo.org> (2022-09-17)
+# petitboot is actively used only on ppc64
+# there are some arm boards too, but it's rare occasion.
+petitboot
```
Comment (mgorny) [inline profiles/arch/base/use.mask:7]: You sure this should be a global flag?

**Review by mgorny — CHANGES_REQUESTED** [review]

**Comment by floppym** [comment]
I don't think this really belongs in the installkernel-systemd-boot package. It's a drop-in snippet for kernel-install, and could be used without having installkernel installed at all.

**Comment by gyakovlev** [comment]
thanks for reviews, ok I'm address @mgorny 's bash comments, close this one and think on how to structure this better and maybe move script to shared org.

**Pull request closed without merging**